OutOfSchool/OutOfSchool.WebApi/Services/Gcp/GcpImagesStorageSynchronizationService.cs

```diff
@@ -0,0 +1,138 @@
+            ILogger<GcpImagesStorageSynchronizationService> logger,
+            IImageFilesStorage imagesStorage,
+            IGcpImagesSyncDataRepository gcpImagesSyncDataRepository)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.imagesStorage = imagesStorage ?? throw new ArgumentNullException(nameof(imagesStorage));
+            this.gcpImagesSyncDataRepository = gcpImagesSyncDataRepository ?? throw new ArgumentNullException(nameof(gcpImagesSyncDataRepository));
+
+            GetAllSyncFunctions = new List<Func<IEnumerable<string>, Task<List<string>>>>
+            {
+                // Entity images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderImagesIds,
+
+                // Entity cover images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectTeacherCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderCoverImagesIds,
+            };
+        }
+
+        private List<Func<IEnumerable<string>, Task<List<string>>>> GetAllSyncFunctions { get; }
+
+        /// <inheritdoc/>
+        public async Task SynchronizeAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                logger.LogDebug("Gcp storage synchronization was started");
+
+                // gcp returns data by local time
+                var dateTime = DateTime.Now.AddHours(-1);
+
+                var listsOfObjects = GetListsOfObjects().ConfigureAwait(false);
+
+                await foreach (var objects in listsOfObjects.WithCancellation(cancellationToken))
+                {
+                    if (objects.Items is null || objects.Items.Count == 0)
+                    {
+                        break;
+                    }
+
+                    var mappedObjects = objects.Items
+                        .Where(x => x.TimeCreated < dateTime)
+                        .Select(x => x.Name)
+                        .ToHashSet();
+
+                    var notAttachedIds = await SynchronizeAll(mappedObjects).ConfigureAwait(false);
+                    await RemoveNotAttachedIds(notAttachedIds).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogCritical(ex, "Gcp storage synchronization was interrupted by error");
+            }
+            finally
+            {
+                logger.LogDebug("Gcp storage synchronization was finished");
+            }
+        }
+
+        private async Task<HashSet<string>> SynchronizeAll(HashSet<string> searchIds)
+        {
+            foreach (var syncFunction in GetAllSyncFunctions)
+            {
+                var syncResult = await syncFunction(searchIds).ConfigureAwait(false);
```

Comment:
Looks pretty confusing, why do we have to wait `sync` function ?

Reply:
`sync` here means synchronizeFunctions not synchronous. See previous comment about `await` if it must be synchronous in executing.

Reply:
That was a hint of renaming this `syncFunction` to something like a `synchronizationFunction` or to the name you already showed in your answer )
Also, if you agree with renaming - please rename property or field `GetAllSyncFunctions` as well